### ISSUE TITLE
Feature: Enable WAP (Write-Audit-Publish) / blue-green deployment

### ISSUE THREAD

**Issue by SoumayaMauthoorMOJ** [dbt-athena/dbt-athena#381]
How do you perform WAP / blue-green deployment with dbt-athena?

With snowflake you can do a database swap: https://discourse.getdbt.com/t/performing-a-blue-green-deploy-of-your-dbt-project-on-snowflake/1349

I was wondering if we could make use of [swap_table()](https://github.com/dbt-athena/dbt-athena/blob/main/dbt/adapters/athena/impl.py#L526) which as per my understanding is currently used for enabling [high availability](https://github.com/dbt-athena/dbt-athena#highly-available-table). 

So the steps would be something like:
```
$ dbt run #making sure ha: false
$ dbt test
$ dbt run-operation swap_database
```

I would be happy to make the changes if people think this is a good idea, but a few pointers to start me off would be very appreciated!

**Comment by nicor88** [comment]
@SoumayaMauthoorMOJ great issue.

I suggest implementing such pattern using Iceberg tables - swap_table is a workaround to guarantee high availability within hive tables (let's exclude it for now).

Here is my suggestion to implement this.
Let's imagine we want to simple WAP for a model called `my_model`.
Here is what needs to be executed sequentially 

* dbt run --select my_model
* dbt test--select my_model
* if tests pass proceed with the public operation
* dbt run-operation publish_table --args '{model: my_model, final_name: my_final_table_name}'

Where `publish_table` is a custom macro that can be implemented in the following way:

```
{%- macro publish_table(reference_model, final_model_name) -%}
  {%- set reference_relation = ref(reference_model) -%}

  {%- if reference_relation is not none -%}

    -- final model relation is the final model name that we use somewhere
    {%- set final_model_relation = adapter.get_relation(
          database=reference_relation.database,
          schema=reference_relation.schema,
          identifier=final_model_name) -%}

    -- if final model exists we rename to a backup version
    {%- if final_model_relation is not none -%}
      {%- set old_relation_bkp = make_temp_relation(final_model_relation, '__wap_bkp') -%}

      {%- if old_relation_bkp is not none -%}
        {%- do drop_relation(old_relation_bkp) -%}
      {%- endif -%}

      {%- do rename_relation(final_model_relation, old_relation_bkp) -%}
    {%- endif -%}

    {%- set taget_relation = api.Relation.create(
      database=reference_relation.database,
      schema=reference_relation.schema, identifier=final_model_name) -%}

    {{ rename_relation(reference_relation, taget_relation) }}
  {%- endif -%}
{%- endmacro -%}
```

The macro above is swapping your model to the final model name that you use for example during reporting.
During the swaping time there are rename operations happening - they latency of this operations is really low (<1 second), but when they happen there could be a little downtime.

Happy to hear your thought on this.

**Comment by SoumayaMauthoorMOJ** [comment]
Hi again @nicor88 will give this a try!

**Comment by nicor88** [comment]
@SoumayaMauthoorMOJ I tried the above and worked for one model - in case you run a selection of models you might want to call the macro for all the WAP models that you have in that run (not all the model needs to follow that pattern)

**Comment by SoumayaMauthoorMOJ** [comment]
Hello @nicor88 sorry got side-tracked  and I still haven't tried your suggestion. I was wondering if/how your approach could work with `dbt build` ? So that each table is independently run, tested, and swapped  before the transformation occurs in the downstream resources. It could be impossible, and then in that case I think would go with the approach you've suggested.

**Comment by nicor88** [comment]
@SoumayaMauthoorMOJ great question again- I don't think that what I propose will work with dbt build.

**Comment by nicor88** [comment]
@SoumayaMauthoorMOJ i will be curious to know how dbt build could work WAP with other adapters (eg snowflake)

**Comment by SoumayaMauthoorMOJ** [comment]
Hey @nicor88 I've been trying to do some research and even posted on dbt (see [comment](https://github.com/dbt-labs/dbt-core/discussions/5687#discussioncomment-6796064) but couldn't get how to modify the internals of `dbt build`. 

However I did wonder why we couldn't follow the same principle as before but with dbt build:
```
$ dbt build
$ dbt run-operation --select result:passed  --state ./target` swap_database
```

This means you could still make use of dbt build's ability to skip models downstream of failed tests. I'm hoping you can then `dbt run-operation` with only models that have passed. Any thoughts?

**Comment by nicor88** [comment]
I think that it should be possible to implement what you suggested - we just have to call the macro `publish_table` for all passed models in the run-operations, which means that the swap will happen only after dbt build - where dbt build checks for upstream source failures, run and then test the models.

Maybe the **publish_table** could be refactored to **publish_tables**, and do the loop inside for all models that were changed. Assuming that you follow a convention for your "published" tables - I believe that should work.

**Comment by SoumayaMauthoorMOJ** [comment]
Perfect I'll look into this when I get back from holidays on 5th September thanks @nicor88 !

**Comment by SoumayaMauthoorMOJ** [comment]
Sorry @nicor88 for the delay on feedback on this issue. I started implementing it and then I remembered that I need be able to re-run failed models. I’m wondering if this approach will be able to handle failed models which have parents which have been successfully refreshed. The model will expect the parent to follow the same naming convention e.g parent_database.parent_table_wap, but will be unable to find it because the table has already been published.

**Comment by SoumayaMauthoorMOJ** [comment]
@nicor88 I’m wondering if this approach might work:
1. Create a glue table copy of the model as a pre-hook (without copying the data)
2. Dbt build
3. Find tables that have failed tests
4. Replace tables that have failed with the glue table copy

This approach should work because with unique naming convention you keep old versions of the table(I’ve tested with iceberg). You can re-run failed models because you’re not applying a different naming convention. The only problem I can think about is that bad data will be visible until it’s replaced. I would also need to find a way to create of copy of the glue table as a pre-hook.

**Comment by nicor88** [comment]
I do believe that might not work, because glue databases cannot renamed...

**Comment by SoumayaMauthoorMOJ** [comment]
@nicor88 I meant something like this to create a duplicate of the glue table:
```
CREATE TABLE <database>.<table>_wap (
  <column details?)
LOCATION 's3://bucket/
TBLPROPERTIES (
  'table_type'='ICEBERG',
  'format'='PARQUET',
  'write_compression'='ZSTD'
);
```
where the parquet files are already present so you don't have to copy the parquet files again.
However I forgot that iceberg tables are different to hive and you can't do this with iceberg tables .

**Comment by nicor88** [comment]
Ah I understood now.

What you suggested could work then, but I suggest a few optimizations:
- make the cloning operation table type agnostic -> it can uses the table type of the table and decide how to backup.
- for hive do not use ONLY metadata - you might want to copy table metadata plus actual data, it's safer. 

In all of this I feel that we are missing some cloning possibilities in the adapter - that will definitely help you a lot on what you want to achieve. I need to think a bit more on how to develop such cloning capabilities, but that can be really beneficial.

**Comment by SoumayaMauthoorMOJ** [comment]
> @SoumayaMauthoorMOJ will you be up for a call to clarify more and understand how we can move on?

Yes I think that would be a good idea! Let's catch up over slack tomorrow? 2pm GMT?

**Comment by SoumayaMauthoorMOJ** [comment]
@nicor88 as discussed I've summarised the new approach on slack https://getdbt.slack.com/archives/CBSQTAPLG/p1694787638360559?thread_ts=1694699328.052999&cid=CBSQTAPLG and waiting for feedback.

**Comment by SoumayaMauthoorMOJ** [comment]
I’ve created a [dbt](https://github.com/moj-analytical-services/jaffle_shop_duckdb_wap) project to demonstrate how to implement WAP on table [materializations](https://docs.getdbt.com/docs/build/materializations) using “dummy” post-test-hooks. It applies the approach to the example [jaffle_shop_db](https://github.com/dbt-labs/jaffle_shop_duckdb) dbt project which uses the [dbt-duckdb](https://github.com/duckdb/dbt-duckdb) adapter.

**Comment by nicor88** [comment]
@SoumayaMauthoorMOJ What you come up with could be really neat if documented in our docs: https://github.com/dbt-athena/dbt-athena.github.io

Also because there are some caveats that are different for athena and duck db I believe - what do you think about opening a pr in our docs to document such a setup? Other users might benefit from that.

**Comment by SoumayaMauthoorMOJ** [comment]
@nicor88 I would be happy to do that but after we go live in 3 weeks - how does that sound?

**Comment by nicor88** [comment]
@SoumayaMauthoorMOJ of course, take your time, there is no rush

**Comment by jelledv** [comment]
Very neat! Any update on this?

**Comment by SoumayaMauthoorMOJ** [comment]
@jelledv not sure if you're still interested but I came up with a wap approach that works for incremental models, but only with Iceberg:

1. Create a ‘wap’ version of your model:
```
--my_table_wap.sql
{{
    config(
        materialized='incremental',
        table_type='iceberg',
    )
}}


SELECT 1 as id

{% if is_incremental() %}

    SELECT 2 as id

{% endif %}
```

2. Create a view model that references the wap model but as of a set [TIMESTAMP](https://docs.aws.amazon.com/athena/latest/ug/querying-iceberg-table-data.html)
```
--my_table.sql
{{
    config(
        materialized='view'
    )
}}

{% set current_timestamp = modules.datetime.datetime.now() %}

SELECT * 
FROM {{ ref('my_table_wap') }}
FOR TIMESTAMP AS OF TIMESTAMP '{{ current_timestamp }} UTC'
```

The view will only get refreshed to use the latest Iceberg table version if the tests passes
